// print_control/WinNumberControl.js
/**
 * Created by w44 on 15-3-18.
 */


var CronJob = require('cron').CronJob;

var xml2js = require('xml2js');
var $ = require('jquery');

var util = require('print_util');
var log = util.log;
var httpUtil = util.httpUtil;
var mongoDBUtil = util.mongoDBUtil;

var cons = require('print_constants');
var nodePlatCons = cons.nodePlatCons;
var moment = require('moment');


var assist = require('print_assist');
var nodePlatAssist = assist.nodePlatAssist;

var WinNumberControl = function () {
};


WinNumberControl.prototype.run = function () {
    var self = this;
    //等待队列消息
    self.winNumberT05Task.start();
};


WinNumberControl.prototype.winNumberT05Task = new CronJob('*/60 * * * * *', function () {
    log.info("...获取11选5开奖号码任务...");
    httpUtil.getT05WinNumber(function (err, html) {
        $(html).find('.chart-bg-qh').each(function () {
            var gameCode = 'T05';
            var code = $(this).text();
            var wNum = $(this).next().text();
            var openTime = moment().format('YYYY-MM-DD hh:mm:ss');
            var winTerm = {
                gameCode: gameCode,
                code: code,
                wNum: wNum,
                openTime: openTime
            };
            mongoDBUtil.db.collection('WinTerms', {safe: true}, function (err, collection) {
                collection.findOne({'gameCode': winTerm.gameCode, 'code': winTerm.code}, function (err, _term) {
                    if (!_term) {
                        //没有这个期次，则记录并通知平台开奖号码
                        var body = {};
                        var term = {};
                        term.gameCode = gameCode;
                        term.code = code + '';
                        term.wNum = wNum;
                        term.status = nodePlatCons.term.status.DRAW;
                        body.term = term;
                        nodePlatAssist.sentP03(body, function (err, backMsgNode) {
                            if (!err) {
                                log.error(backMsgNode.body);
                                var _body = JSON.parse(backMsgNode.body);
                                if (_body.repCode == '0000') {
                                    collection.insert(winTerm, function (err, data) {
                                        log.error('已经成功通知平台，并记录本次开奖期次');
                                        log.error(data);
                                    });
                                }
                                ;
                            }
                        });
                    } else {
                        log.info(JSON.stringify(_term) + '已返给平台开奖!');
                    }
                });
            });


        });


        parser.parseString(resData, function (err, data) {
            var gameCode = 'T05';
            var code = data.root.lastNo[0];
            var wNum = data.root.lastRes[0];
            var reg = new RegExp("\\+", "g"); //创建正则RegExp对象
            wNum = wNum.replace(reg, ',');
            var openTime = data.root.time[0];
            var winTerm = {
                gameCode: gameCode,
                code: code,
                wNum: wNum,
                openTime: openTime
            };
            mongoDBUtil.db.collection('WinTerms', {safe: true}, function (err, collection) {
                collection.findOne({'gameCode': winTerm.gameCode, 'code': winTerm.code}, function (err, _term) {
                    if (!_term) {
                        //没有这个期次，则记录并通知平台开奖号码
                        var body = {};
                        var term = {};
                        term.gameCode = gameCode;
                        term.code = code + '';
                        term.wNum = wNum;
                        term.status = nodePlatCons.term.status.DRAW;
                        body.term = term;
                        nodePlatAssist.sentP03(body, function (err, backMsgNode) {
                            if (!err) {
                                log.error(backMsgNode.body);
                                var _body = JSON.parse(backMsgNode.body);
                                if (_body.repCode == '0000') {
                                    collection.insert(winTerm, function (err, data) {
                                        log.error('已经成功通知平台，并记录本次开奖期次');
                                        log.error(data);
                                    });
                                }
                                ;
                            }
                        });
                    } else {
                        log.info(JSON.stringify(_term) + '已返给平台开奖!');
                    }
                });
            });
        });
    });
}, null, false, 'Asia/Chongqing');


var winNumberControl = new WinNumberControl();

module.exports = winNumberControl;
