// print_config/Prop.js
/**
 * Created by w44 on 15-1-12.
 */


var exports={};

var target='dev';
var argv = process.argv;
var kvs = {};
for(var key in argv)
{
    if(key > 1)
    {
        var kv = argv[key].split("=");
        kvs[kv[0]] = kv[1];
    }
}
if(kvs.target)
{
    target = kvs.target;
}
var HOST = '123.57.166.122';
var PORT = 16777;

if(target=='run'){
    exports.dbName='printer';
    exports.mongodbUrl='localhost';
    exports.mongodbPort=27017;

    HOST = '123.57.166.122';
    PORT = 16777;

}else if(target=='dev'){
    exports.dbName='printer';

    exports.mongodbUrl='123.57.166.122';
    exports.mongodbPort=27017;
    //HOST = '101.200.232.194';
	//HOST = '192.168.200.150'

	HOST = 'localhsot'
    PORT = 16777;
}else if(target=='test'){
    exports.dbName='printer';
    exports.mongodbUrl='localhost';
    exports.mongodbPort=27017;
    HOST = '123.57.166.122';
    PORT = 16777;
}
exports.HOST = HOST;
exports.PORT = PORT;
   
var node_plat_https={
    'devOpt': {
        hostname: '123.57.166.122',
        port: 9088,
        path: '/mcp-filter/main/interface.htm',
        method: 'POST'
    },
    'runOpt': {
        hostname: 'localhost',
        port: 9088,
        path: '/mcp-filter/main/interface.htm',
        method: 'POST'
    }
};


if(target == 'dev')
{
    exports.node_plat_opt = node_plat_https.devOpt;
}
else if(target == 'run')
{
    exports.node_plat_opt = node_plat_https.runOpt;
}


module.exports = exports;
