// print_assist/NodePlatAssist.js
/**
 * Created by w44 on 15-1-12.
 */

var moment = require('moment');
var util = require('print_util');
var httpUtil = util.httpUtil;
var log=util.log;

var cons = require('print_constants');
var nodePlatCons = cons.nodePlatCons;
var dataTransCons = cons.dataTransCons;


var NodePlatAssist = function () {
};


NodePlatAssist.prototype.sent = function (cmd, bodyStr, cb) {
        httpUtil.sendToNodePlat(cmd, bodyStr, cb);
};

NodePlatAssist.prototype.sentP01 = function (cb) {
    var self = this;
    var body = {};
    self.sent(nodePlatCons.command.queryOrder, body, function (err, backMsgNode) {
        if (err) {
            cb(err, null);
        }
        else {
            cb(null, backMsgNode);
        }
    });
};


NodePlatAssist.prototype.sentP02 = function (ticket, cb) {
    log.error(ticket.id+'出票成功，正在返回给平台结果');
    var self = this;
    var body = {};
    //根据模式选择发送的body
    var rst = [];
    body.rst = rst;
    var _ticket = {
        id: parseFloat(ticket.id),
        status: nodePlatCons.ticket.status.success,
        province: 'bj',
        seq: ticket.ticketSeq,
        terminal: ticket.terminalId,
        printTime: moment().format('YYYY-MM-DD HH:mm:ss'),
        passw:ticket.ticketPwd2,
        rNumber: ''
    };
    //todo 在这里加上对竞彩的判断
    if (ticket.gameCode == 'T51') {
        var results = ticket.metaTicket.split(new RegExp('周'));
        if(!results){
            cb('竞彩ticket.metaTicket.match(regExp)为空', null);
            return
        };
        var ticketCache = {};
        var numberArray = ticket.numbers.split(";")
        for(var i = 0; i < numberArray.length; i++){
            var matchCode = numberArray[i].split("|")[1];
            var match = matchCode.substr(matchCode.length - 4);
            ticketCache[match] = matchCode;
        }
        var resultArray = new Array();
        for(var i = 0; i< results.length; i++){
            if(i != 0){
                var arr = results[i].split('\n\n');
                var day = arr[0].split("\n")[0];
                if(arr[0].indexOf(':') > 0){
                    var temp = arr[0].substr(0, arr[0].indexOf(':'));
                    day = temp.replace('日', 7).replace('一', 1).replace('二', 2).replace('三', 3).replace('四', 4).replace('五', 5).replace('六', 6).replace(/[^0-9]/g,'');
                }else{
                    day.replace('日', 7).replace('一', 1).replace('二', 2).replace('三', 3).replace('四', 4).replace('五', 5).replace('六', 6).replace(/[^0-9]/g, '');
                }
                var resultTemp = arr[1].replace(/\n/g,'');
                var pl = "";
                var pType = ticket.playTypeCode;
                if (resultTemp.indexOf('(') >= 0 && resultTemp.indexOf(":") > 0) {
                    pType = "03";
                    pl = resultTemp.replace(/\s+/g, '').replace(/\+/g, ',').replace(/元/g, '').replace(/\:/g, '').replace(/\(/g, '').replace(/\)/g, '');
                }
                if (resultTemp.indexOf('(') >= 0 && resultTemp.indexOf('其它') > 0) {
                    pType = "03";
                    pl = resultTemp.replace(/\s+/g, '').replace(/\+/g, ',').replace(/元/g, '').replace(/\:/g, '').replace(/\(/g, '').replace(/\)/g, '').replace(/胜其它/g, '90').replace(/平其它/g, '99').replace(/负其它/g, '09');
                }
                if (resultTemp.indexOf('(') >= 0 && resultTemp.indexOf(":") < 0 && resultTemp.indexOf('其它') < 0) {
                    pType = "04";
                    pl = resultTemp.replace(/\+\)/g, '').replace(/\+/g, ',').replace(/元/g, '').replace(/\(/g, '').replace(/\)/g, '');
                }
                if (resultTemp.indexOf('(') < 0) {
                    if (arr[0].indexOf('让球') > 0) {
                        pType = "01";
                    } else if (arr[0].indexOf('半全场') > 0) {
                        pType = "05";
                    } else if (arr[0].indexOf('胜平负') > 0) {
                        pType = "02";
                    }
                    pl = resultTemp.replace(/\s+/g, '').replace(/\+/g, ',').replace(/元/g, '').replace(/负/g, 0).replace(/平/g, 1).replace(/胜/g, 3);
                }
                resultArray.push(pType + '|' + ticketCache[day] + '|' + pl);
            }
        }
        _ticket.rNumber = resultArray.join(';')
    }


    if (ticket.gameCode == 'T52') {
        var results = ticket.metaTicket.split(new RegExp('周'));
        if(!results){
            cb('竞彩ticket.metaTicket.match(regExp)为空', null);
            return
        };
        var ticketCache = {};
        var pTypeCache = {};
        var numberArray = ticket.numbers.split(";")
        for(var i = 0; i < numberArray.length; i++){
            var matchCode = numberArray[i].split("|")[1];
            var match = matchCode.substr(matchCode.length - 4);
            ticketCache[match] = matchCode;
            var pCode=numberArray[i].split("|")[0];
            pTypeCache[match]=pCode;
        }
        var resultArray = new Array();
        for(var i = 0; i< results.length; i++){
            if(i != 0){
                var arr = results[i].split('\n\n');
                var day = arr[0].split("\n")[0];
                var yushe;
                if( day.indexOf(':')>0 && (day.indexOf('让分')>0 || day.indexOf('总分')>0)) {
                    yushe=day.split(':')[1];
                    if(yushe.indexOf('主') >= 0){
                        yushe = yushe.split('主')[1];
                        yushe = '('+yushe+')';
                    }else if (yushe.indexOf('客') >= 0){
                        yushe = yushe.split('客')[1];
                        yushe = '('+yushe+')';
                    }else {
                        yushe = '('+yushe+')';
                    }
                }
                if(arr[0].indexOf(':') > 0){
                    var temp = arr[0].substr(0, arr[0].indexOf(':'));
                    day = temp.replace('日', 7).replace('一', 1).replace('二', 2).replace('三', 3).replace('四', 4).replace('五', 5).replace('六', 6).replace(/[^0-9]/g,'');
                }else{
                    day.replace('日', 7).replace('一', 1).replace('二', 2).replace('三', 3).replace('四', 4).replace('五', 5).replace('六', 6).replace(/[^0-9]/g, '');
                }
                var resultTemp = arr[1].replace(/\n/g,'');
                var pl = "";

                pl = resultTemp.replace(/\s+/g, '').replace(/\+/g, ',').replace(/元/g, '').replace(/负/g, 2).replace(/胜/g, 1).replace(/大/g, 1).replace(/小/g, 2);

                var jclqTrans = dataTransCons.jclqSfc;
                for(var key in jclqTrans){
                    pl = pl.replace(key,jclqTrans[key]);
                }
                if(pTypeCache[day] == '01' || pTypeCache[day] == '04'){
                    pl = pl.replace(/@/g,yushe+'@');
                }
                resultArray.push(pTypeCache[day]+ '|' + ticketCache[day] + '|' + pl);
            }
        }
        _ticket.rNumber = resultArray.join(';')
    }


    rst.push(_ticket);
    var id = body.rst[0].id;
    var length = 32 - (id + '').length;
    for (var i = 0; i < length; i++) {
        id = '0' + id;
    };
    log.error(JSON.stringify(body));
    self.sent(nodePlatCons.command.ticketResponse, body, function (err, backMsgNode) {
        log.error('P02响应:'+JSON.stringify(backMsgNode));
        if (err) {
            cb(err, null);
        }
        else {
            cb(null, backMsgNode);
        }
    });
};

NodePlatAssist.prototype.sentFailed = function (ticket, cb) {
    var self = this;
    var body = {};
    //根据模式选择发送的body
    var rst = [];
    body.rst = rst;
    var _ticket = {
        id: parseFloat(ticket.id),
        status: nodePlatCons.ticket.status.fail,
        province: 'bj',
        seq: ticket.id,
        terminal: ticket.terminalId,
        rNumber: ''
    };
    //todo 在这里加上对竞彩的判断
//    if (ticket.gameCode == 'T51') {
//        var regExp = new RegExp('周.*[0-9]{3}\\n主队:.*Vs\\s客队:.*\\n\\n.*', 'g');
//        var results = ticket.metaTicket.match(regExp);
//        var now = moment().format('YYYYMMDD');
//        if(!results){
//            cb('竞彩ticket.metaTicket.match(regExp)为空', null);
//            return;
//        };
//        results.forEach(function (item) {
//            var arr = item.split('\n');
//            var day = arr[0].replace('周日', 7).replace('周一', 1).replace('周二', 2).replace('周三', 4).replace('周四', 4).replace('周五', 5).replace('周六', 6);
//            var pl = arr[3].replace(/\s+/g, '').replace(/\+/g, ',').replace(/元/g, '').replace('负', 0).replace('平', 1).replace('胜', 2);
//            _ticket.rNumber += ticket.playTypeCode + '|' + now + day + '|' + pl + ';'
//        });
//    }

    rst.push(_ticket);
    var id = body.rst[0].id;
    var length = 32 - (id + '').length;
    for (var i = 0; i < length; i++) {
        id = '0' + id;
    };
    log.error("错误票返回平台发送:"+JSON.stringify(body));
    self.sent(nodePlatCons.command.ticketResponse, body, function (err, backMsgNode) {
        log.error("错误票返回平台响应:"+JSON.stringify(backMsgNode));
        if (err) {
            cb(err, null);
        }
        else {
            cb(null, backMsgNode);
        }
    });
};

NodePlatAssist.prototype.sentCQ01 = function (bodyNode, cb) {
    var self = this;
    self.sent(nodePlatCons.command.queryTerm, bodyNode, function (err, backMsgNode) {
        if (err) {
            cb(err, null);
        }
        else {
            cb(null, backMsgNode);
        }
    });
};


NodePlatAssist.prototype.sentP03 = function (body,cb) {
    var self = this;
    self.sent(nodePlatCons.command.sendNumbers, body, function (err, backMsgNode) {
        if (err) {
            cb(err, null);
        }
        else {
            cb(null, backMsgNode);
        }
    });
};

var nodePlatAssist = new NodePlatAssist();

module.exports = nodePlatAssist;
